### PR TITLE
PR: Handle paths for recent files on a different mount than the project root path

### LINKED ISSUE #10045: ValueError when closing project on USB stick
### What steps will reproduce the problem?

1. Open project from USB stick with Spyder 3
2. Update to Spyder 4 beta
3. Try closing project. Results in an error.

### Traceback
```python-traceback
  File "C:\Users\jmgro\Anaconda3\lib\site-packages\spyder\plugins\projects\plugin.py", line 342, in close_project
    self.main.editor.get_open_filenames())
  File "C:\Users\jmgro\Anaconda3\lib\site-packages\spyder\plugins\projects\plugin.py", line 427, in set_project_filenames
    self.current_active_project.set_recent_files(recent_files)
  File "C:\Users\jmgro\Anaconda3\lib\site-packages\spyder\plugins\projects\api.py", line 76, in set_recent_files
    for recent_file in recent_files]
  File "C:\Users\jmgro\Anaconda3\lib\site-packages\spyder\plugins\projects\api.py", line 76, in <listcomp>
    for recent_file in recent_files]
  File "C:\Users\jmgro\Anaconda3\lib\ntpath.py", line 562, in relpath
    path_drive, start_drive))
ValueError: path is on mount 'C:', start on mount 'E:'
```

## Versions

* Spyder version: 4.0.0b4 
* Python version: 3.7.3
* Qt version: 5.9.6
* PyQt5 version: 5.9.2
* Operating System: Windows 10

### Dependencies

```
pygments >=2.0    :  2.4.2 (OK)
qdarkstyle >=2.6.4:  2.7 (OK)
sphinx >=0.6.6    :  2.1.2 (OK)
pyls >=0.27.0     :  0.28.0 (OK)
nbconvert >=4.0   :  5.5.0 (OK)
pandas >=0.13.1   :  0.24.2 (OK)
numpy >=1.7       :  1.16.4 (OK)
sympy >=0.7.3     :  1.4 (OK)
cython >=0.21     :  0.29.12 (OK)
qtconsole >=4.5.2 :  4.5.3 (OK)
IPython >=4.0     :  7.6.1 (OK)
matplotlib >=2.0.0:  3.1.0 (OK)
pylint >=0.25     :  2.3.1 (OK)
```

### PR DESCRIPTION
**Pull request by dalthviz**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Handle recent files saving in cases where paths doesn't have a relative path to the project root path

### Issue(s) Resolved

Fixes #10045

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: dalthviz

### CODE REVIEW

File: spyder/plugins/projects/projecttypes/__init__.py [hunk 1]
```diff
@@ -75,11 +78,6 @@ def get_recent_files(self):
         recent_files = [recent_file if os.path.isabs(recent_file)
                         else os.path.join(self.root_path, recent_file)
                         for recent_file in recent_files]
-
-        for recent_file in recent_files[:]:
-            if not os.path.isfile(recent_file):
-                recent_files.remove(recent_file)
-
```
Comment (ccordoba12) [inline spyder/plugins/projects/projecttypes/__init__.py]: Why did you remove these lines? Without these lines `get_recent_files` does nothing.

Comment (dalthviz) [reply spyder/plugins/projects/projecttypes/__init__.py]: I made this change (removing the validation of the path being a file) to check the tests (`test_set_get_project_filenames_when_closing` and `test_set_get_project_filenames_when_switching`). Seems like the existing tests were passing because of the mutability property of a list (the `set_recent_files` method was changing directly the `opened_files` list that at the end is the object passed by parameter, so the [assert](https://github.com/spyder-ide/spyder/blob/aafb1c21f6e8c58997e43ccfc8cf8d06e2feb9c6/spyder/plugins/projects/tests/test_plugin.py#L197) was checking the return with a list changed by the `set_recent_files` method (the recent files in the tests always were an empty list, since the paths we are testing here aren't actually created).

However, probably we need here to change the tests to create the files and add again the check for the path being a file. So this is not ready yet :)

Comment (ccordoba12) [reply spyder/plugins/projects/projecttypes/__init__.py]: > So this is not ready yet :)

@dalthviz, then please don't forget to add WIP to the PR title.

Comment (dalthviz) [reply spyder/plugins/projects/projecttypes/__init__.py]: Yep :+1:

**Comment by goanpeca** [comment]
@dalthviz just to be sure, could you run the tests reverting the fix on code and making sure it fails? 

This is to ensure that we test the regression.

File: spyder/plugins/projects/tests/test_plugin.py [hunk 2]
```diff
@@ -175,18 +175,52 @@ def test_open_project_uses_visible_config(projects, tmpdir, value):
     assert projects.dockwidget.isVisible() == value
 
 
-def test_set_get_project_filenames_when_closing(create_projects, tmpdir):
+def test_set_get_project_filenames_when_closing_no_files(create_projects,
+                                                         tmpdir):
```
Comment (ccordoba12) [inline spyder/plugins/projects/tests/test_plugin.py]: I don't understand how this new test checks a regression for issue #10045

Comment (dalthviz) [reply spyder/plugins/projects/tests/test_plugin.py]: This is to check the old unwanted behavior of the `set_recent_files` where the method changed the `opened_files` list (because the old method modified the parameter as explained above) making the old tests pass. If you try to run this test with the old implementation of the `set_recent_files` method it fails.

Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: But is not the issue related to using relative paths? Or what's the real problem here?

Comment (dalthviz) [reply spyder/plugins/projects/tests/test_plugin.py]: Yes, the main issue is related to the root path of a project being in a different mount path/drive than the files currently open. So, for example working on Windows, the project is on `C` but the files are on `D`, so no relative path exist between the project root path and the file. (that way I added the `try-except`).

However, while doing the fix, I discover that the tests which were supposed to check that the recent files elements worked on projects were kind of faulty, partially because of the way `set_recent_files` was implemented (the variable `opened_files` use to do the assert in the test, was modified by the `set_recent_files` method itself).

Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: Ok, thanks @dalthviz for the thorough explanation!
